### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by sveisvei**
* Breaking: requires `appName` to be provided
* register client with strategynames at /client/register on startup
* report 60 seconds buckets of metrics to /client/metrics on unleash server
* exposing metric events

cc @ivarconr @SimenB

### CODE REVIEW

File: src/metrics.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    count (name: string, enabled: boolean) : void {
+        if (!this.store[name]) {
+            this.store[name] = {
+                yes: [],
+                no: [],
+            };
+        }
+        this.store[name][enabled ? 'yes' : 'no'].push({
+            time: new Date(),
```
Comment (SimenB) [inline src/metrics.ts]: `Date.now`?

File: src/metrics.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+
+    count (name: string, enabled: boolean) : void {
+        if (!this.bucket.toggles[name]) {
+            this.bucket.toggles[name] = {
+                yes: 0,
+                no: 0,
+            };
+        }
+        this.bucket.toggles[name][enabled ? 'yes' : 'no'] += 1;
```
Comment (SimenB) [inline src/metrics.ts]: `++`?

File: src/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+            if (res.statusCode !== 200) {
+                this.emit('warn', `${url} returning ${res.statusCode}`);
+                return;
+            }
+        });
+    }
+
+
+    count (name: string, enabled: boolean) : void {
```
Comment (SimenB) [inline src/metrics.ts]: I'd love to see this emit the fact it counted 😄

File: src/unleash.ts [hunk 4]
```diff
@@ -34,18 +42,51 @@ export class Unleash extends EventEmitter {
...
+        this.repository.on('error', (err) => {
+            err.message = `Unleash Repository error: ${err.message}`;
+            this.emit('error', err);
+        });
+
+        this.repository.on('warn', (msg) => {
+            this.emit('warn', msg);
+        });
+
+        this.metrics = new Metrics({
```
Comment (ivarconr) [inline src/unleash.ts]: we might want a flag to disable metrics in case it prove to be some perf issues?

Comment (sveisvei) [inline src/unleash.ts]: inc flag

File: src/metrics.ts (hunk 3 above)
Comment (sveisvei) [inline src/metrics.ts]: Whats the usecase? :)

Comment (SimenB) [inline src/metrics.ts]: https://github.com/Unleash/unleash-client-node/pull/20#pullrequestreview-6004982

> so we can add prometheus on the consumer side

File: src/client.ts [hunk 5]
```diff
@@ -55,6 +55,10 @@ export default class UnleashClient {
             return false;
         }
 
+        if (feature.strategies.length === 0) {
```
Comment (sveisvei) [inline src/client.ts]: This is somewhat unreleasted @ivarconr, but should be default behaviour?

**Comment by SimenB** [comment]
Why does the client report to the server, instead of the server asking for metrics on demand? Sorta like prometheus does it, you just say "hey, I've got metrics!" and the server then polls it when it wants to.

File: src/metrics.ts [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        if (this.disabled) {
+            return false;
+        }
+        if (!this.bucket.toggles[name]) {
+            this.bucket.toggles[name] = {
+                yes: 0,
+                no: 0,
+            };
+        }
+        this.bucket.toggles[name][enabled ? 'yes' : 'no'] ++;
```
Comment (SimenB) [inline src/metrics.ts]: space before `++`? Is that our eslint rule?

File: src/request.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import * as request from 'request';
```
Comment (SimenB) [inline src/request.ts]: Why not a promise based one? As we're close to async await, that's cleaner, IMO

File: src/unleash.ts [hunk 8]
```diff
@@ -9,8 +10,12 @@ import { EventEmitter } from 'events';
 const BACKUP_PATH: string = tmpdir();
 
 export interface UnleashConfig {
+    appName?: string,
```
Comment (SimenB) [inline src/unleash.ts]: should it be `?` if you throw if it's missing?

File: src/unleash.ts [hunk 9]
```diff
@@ -34,32 +44,67 @@ export class Unleash extends EventEmitter {
...
+        this.metrics = new Metrics({
+            disableMetrics,
+            appName,
+            instanceId,
+            strategies: strategies.map((strategy: Strategy) => strategy.name),
+            metricsInterval,
+            url
+        });
+
+        this.metrics.on('error', (err) => {
```
Comment (SimenB) [inline src/unleash.ts]: still missing an `this.metrics.on(''count')` so we can also chuck this in the consumers of the client. Seems like only I want that though

File: test/metrics.network.test.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import test from 'ava';
+import Metrics from '../lib/metrics';
+import nock from 'nock';
+
+nock.disableNetConnect();
```
Comment (SimenB) [inline test/metrics.network.test.js]: should be in a `before`, and removed in an `after`, or your tests have side effects

File: test/metrics.test.js [hunk 11]
```diff
@@ -0,0 +1,197 @@
+import test from 'ava';
+import Metrics from '../lib/metrics';
+import nock from 'nock';
+
+let counter = 1;
+const getUrl = () => `http://test${counter++}.app/`;
```
Comment (SimenB) [inline test/metrics.test.js]: side efects? not very ava-ish

File: test/unleash.test.js [hunk 12]
```diff
@@ -6,6 +6,9 @@ import mkdirp from 'mkdirp';
 
 import { Unleash } from '../lib/unleash';
 
+let counter = 1;
+const getUrl = () => `http://test2${counter++}.app/`;
```
Comment (SimenB) [inline test/unleash.test.js]: again, not a fan of this side effect

**Comment by SimenB** [comment]
Some nits, but overall LGTM!

File: test/metrics.network.test.js (hunk 10 above)
Comment (sveisvei) [inline test/metrics.network.test.js]: Indeed

File: src/metrics.ts (hunk 3 above)
Comment (sveisvei) [inline src/metrics.ts]: Sure!

File: src/unleash.ts (hunk 9 above)
Comment (sveisvei) [inline src/unleash.ts]: Will add

File: test/metrics.test.js (hunk 11 above)
Comment (sveisvei) [inline test/metrics.test.js]: I dont want nock to interfere cross tests, could do random instead()?

**Comment by sveisvei** [comment]
Merged this. 

Im not 100% sure on the event names from the metrics @SimenB , if any have better suggestions, let me know.

File: src/metrics.ts (hunk 6 above)
Comment (sveisvei) [inline src/metrics.ts]: This is typescript. I have not hooked up tslint atm.

File: src/request.ts (hunk 7 above)
Comment (sveisvei) [inline src/request.ts]: Did test, but code did not look better with promises on the request part.

File: src/unleash.ts (hunk 8 above)
Comment (sveisvei) [inline src/unleash.ts]: Good catch!

**Comment by ivarconr** [comment]
@SimenB we can not do what prometheus do for multiple reasons:

- we can not enforce all users to expose a port for us (we might move in direction of unleash hosted at some point)
- not all clients are using HTTP (also thrift services are using unleash)
- It would be a total redesign. I am more in favor of moving in direction of websockets or similar to mininmize the load.

**Pull request merged**